[c10d] monitored_barrier: make other ranks fail when one rank doesn't reach the barrier

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55265 [c10d] Log API usage of monitored barrier
* **#55197 [c10d] monitored_barrier: make other ranks fail when one rank doesn't reach the barrier**

From initial user feedback, one unexpected difference between monitored_barrier impl and barrier is the "all or nothing" semantics.

In barrier, all ranks pass or they all fail. With monitored barrier however, if rank 1 is healthy, it will respond to both send and recv from rank 0, but rank 0 can later fail because rank 2 is stuck. In this case, rank 1 will move forward out of the barrier.

This change makes it so that if a rank fails in monitored barrier, all other ranks in monitored barrier will also fail. It does so by the following process, similar to acknowledgements:

Nonzero ranks call send()
Nonzero ranks call recv()

Rank 0 calls recv(), if this succeeds, rank 0 has acknowledged rank N as healthy
Once all ranks are acknowledged as healthy:
Rank 0 calls send() to all nonzero ranks to unblock them

Modified unittests to ensure the all or nothing failure behavior

Differential Revision: [D27523060](https://our.internmc.facebook.com/intern/diff/D27523060/)

rohan-varma committed:
[c10d] monitored_barrier: ensure all ranks pass or none do

From initial user feedback, one unexpected difference between monitored_barrier impl and barrier is the "all or nothing" semantics.

In barrier, all ranks pass or they all fail. With monitored barrier however, if rank 1 is healthy, it will respond to both send and recv from rank 0, but rank 0 can later fail because rank 2 is stuck. In this case, rank 1 will move forward out of the barrier.

This change makes it so that if a rank fails in monitored barrier, all other ranks in monitored barrier will also fail. It does so by the following process, similar to acknowledgements:

Nonzero ranks call send()
Nonzero ranks call recv()

Rank 0 calls recv(), if this succeeds, rank 0 has acknowledged rank N as healthy
Once all ranks are acknowledged as healthy:
Rank 0 calls send() to all nonzero ranks to unblock them

Modified unittests to ensure the all or nothing failure behavior

Differential Revision: [D27523060](https://our.internmc.facebook.com/intern/diff/D27523060/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] monitored_barrier: ensure all ranks pass or none do"


From initial user feedback, one unexpected difference between monitored_barrier impl and barrier is the "all or nothing" semantics.

In barrier, all ranks pass or they all fail. With monitored barrier however, if rank 1 is healthy, it will respond to both send and recv from rank 0, but rank 0 can later fail because rank 2 is stuck. In this case, rank 1 will move forward out of the barrier.

This change makes it so that if a rank fails in monitored barrier, all other ranks in monitored barrier will also fail. It does so by the following process, similar to acknowledgements:

Nonzero ranks call send()
Nonzero ranks call recv()

Rank 0 calls recv(), if this succeeds, rank 0 has acknowledged rank N as healthy
Once all ranks are acknowledged as healthy:
Rank 0 calls send() to all nonzero ranks to unblock them

Modified unittests to ensure the all or nothing failure behavior

Differential Revision: [D27523060](https://our.internmc.facebook.com/intern/diff/D27523060/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] monitored_barrier: ensure all ranks pass or none do"


From initial user feedback, one unexpected difference between monitored_barrier impl and barrier is the "all or nothing" semantics.

In barrier, all ranks pass or they all fail. With monitored barrier however, if rank 1 is healthy, it will respond to both send and recv from rank 0, but rank 0 can later fail because rank 2 is stuck. In this case, rank 1 will move forward out of the barrier.

This change makes it so that if a rank fails in monitored barrier, all other ranks in monitored barrier will also fail. It does so by the following process, similar to acknowledgements:

Nonzero ranks call send()
Nonzero ranks call recv()

Rank 0 calls recv(), if this succeeds, rank 0 has acknowledged rank N as healthy
Once all ranks are acknowledged as healthy:
Rank 0 calls send() to all nonzero ranks to unblock them

Modified unittests to ensure the all or nothing failure behavior

Differential Revision: [D27523060](https://our.internmc.facebook.com/intern/diff/D27523060/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] monitored_barrier: ensure all ranks pass or none do"


From initial user feedback, one unexpected difference between monitored_barrier impl and barrier is the "all or nothing" semantics.

In barrier, all ranks pass or they all fail. With monitored barrier however, if rank 1 is healthy, it will respond to both send and recv from rank 0, but rank 0 can later fail because rank 2 is stuck. In this case, rank 1 will move forward out of the barrier.

This change makes it so that if a rank fails in monitored barrier, all other ranks in monitored barrier will also fail. It does so by the following process, similar to acknowledgements:

Nonzero ranks call send()
Nonzero ranks call recv()

Rank 0 calls recv(), if this succeeds, rank 0 has acknowledged rank N as healthy
Once all ranks are acknowledged as healthy:
Rank 0 calls send() to all nonzero ranks to unblock them

Modified unittests to ensure the all or nothing failure behavior

Differential Revision: [D27523060](https://our.internmc.facebook.com/intern/diff/D27523060/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] monitored_barrier: ensure all ranks pass or none do"


From initial user feedback, one unexpected difference between monitored_barrier impl and barrier is the "all or nothing" semantics.

In barrier, all ranks pass or they all fail. With monitored barrier however, if rank 1 is healthy, it will respond to both send and recv from rank 0, but rank 0 can later fail because rank 2 is stuck. In this case, rank 1 will move forward out of the barrier.

This change makes it so that if a rank fails in monitored barrier, all other ranks in monitored barrier will also fail. It does so by the following process, similar to acknowledgements:

Nonzero ranks call send()
Nonzero ranks call recv()

Rank 0 calls recv(), if this succeeds, rank 0 has acknowledged rank N as healthy
Once all ranks are acknowledged as healthy:
Rank 0 calls send() to all nonzero ranks to unblock them

Modified unittests to ensure the all or nothing failure behavior

Differential Revision: [D27523060](https://our.internmc.facebook.com/intern/diff/D27523060/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] monitored_barrier: ensure all ranks pass or none do"


From initial user feedback, one unexpected difference between monitored_barrier impl and barrier is the "all or nothing" semantics.

In barrier, all ranks pass or they all fail. With monitored barrier however, if rank 1 is healthy, it will respond to both send and recv from rank 0, but rank 0 can later fail because rank 2 is stuck. In this case, rank 1 will move forward out of the barrier.

This change makes it so that if a rank fails in monitored barrier, all other ranks in monitored barrier will also fail. It does so by the following process, similar to acknowledgements:

Nonzero ranks call send()
Nonzero ranks call recv()

Rank 0 calls recv(), if this succeeds, rank 0 has acknowledged rank N as healthy
Once all ranks are acknowledged as healthy:
Rank 0 calls send() to all nonzero ranks to unblock them

Modified unittests to ensure the all or nothing failure behavior

Differential Revision: [D27523060](https://our.internmc.facebook.com/intern/diff/D27523060/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[c10d] monitored_barrier: ensure all ranks pass or none do"


From initial user feedback, one unexpected difference between monitored_barrier impl and barrier is the "all or nothing" semantics.

In barrier, all ranks pass or they all fail. With monitored barrier however, if rank 1 is healthy, it will respond to both send and recv from rank 0, but rank 0 can later fail because rank 2 is stuck. In this case, rank 1 will move forward out of the barrier.

This change makes it so that if a rank fails in monitored barrier, all other ranks in monitored barrier will also fail. It does so by the following process, similar to acknowledgements:

Nonzero ranks call send()
Nonzero ranks call recv()

Rank 0 calls recv(), if this succeeds, rank 0 has acknowledged rank N as healthy
Once all ranks are acknowledged as healthy:
Rank 0 calls send() to all nonzero ranks to unblock them

Modified unittests to ensure the all or nothing failure behavior

Differential Revision: [D27523060](https://our.internmc.facebook.com/intern/diff/D27523060/)

[ghstack-poisoned]

zhaojuanmao commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2738,94 +2738,109 @@ void ProcessGroupGloo::monitoredBarrier(
     auto recvWork = recv(commTensor, 0, t2);
-    sendWork->wait();
-    recvWork->wait();
+    try {
```
How about for the following case?

rank 0 received msgs from all ranks first, then send msgs to rank 1 and rank 2, rank1 recvs the msg and will exit monitoredTimeout successfully first, later on rank2 failed to recv the msg, so rank2 will throw errors.  

in this case,  not all ranks failed in monitored barrier, right?

rohan-varma replied:
In general, if rank != 0, and there is a failure, it will fail before monitored barrier which is the main use case we aim to support, i.e. it will fail before `send()`. In this case, the PR makes it so that all ranks pass or none of them do. 

If all ranks are able to process `send()` but then one fails in `recv()`, it implies that the rank did indeed reach `monitored_barrier` but then somehow errored during it (closed connection, etc). In this case I think it's fine for rank 2 to throw errors (which will be reported by 0) and rank 1 to continue. 

There also doesn't seem to be a great way to solve this case in general, since any subsequent communication could also fail in the same way.

zhaojuanmao replied:
that makes sense for major use cases of the monitored barrier, but the doc said 'Note that this collective is all or nothing, in that all ranks will pass the monitored barrier or none of them will.'==> do you want to refactor the doc a little bit? otherwise, the codes look good to me. thanks for the change!

rohan-varma replied:
That's a good point, will update the doc appropriately as it could be misleading in the case you pointed out.

rohan-varma committed:
Update on "[c10d] monitored_barrier: make other ranks fail when one rank doesn't reach the barrier"


From initial user feedback, one unexpected difference between monitored_barrier impl and barrier is the "all or nothing" semantics.

In barrier, all ranks pass or they all fail. With monitored barrier however, if rank 1 is healthy, it will respond to both send and recv from rank 0, but rank 0 can later fail because rank 2 is stuck. In this case, rank 1 will move forward out of the barrier.

This change makes it so that if a rank fails in monitored barrier, all other ranks in monitored barrier will also fail. It does so by the following process, similar to acknowledgements:

Nonzero ranks call send()
Nonzero ranks call recv()

Rank 0 calls recv(), if this succeeds, rank 0 has acknowledged rank N as healthy
Once all ranks are acknowledged as healthy:
Rank 0 calls send() to all nonzero ranks to unblock them

Modified unittests to ensure the all or nothing failure behavior

Differential Revision: [D27523060](https://our.internmc.facebook.com/intern/diff/D27523060/)

[ghstack-poisoned]
